File: src/parser/luhn.ts
import { normalizePan } from './normalizePan';

export function isValidLuhn(value: string | null | undefined): boolean {
  const pan = normalizePan(value);

  if (pan.length < 12 || pan.length > 19) {
    return false;
  }

  let sum = 0;
  let shouldDouble = false;

  for (let index = pan.length - 1; index >= 0; index -= 1) {
    let digit = Number(pan[index]);

    if (Number.isNaN(digit)) {
      return false;
    }

    if (shouldDouble) {
      digit *= 2;
      if (digit > 9) {
        digit -= 9;
      }
    }

    sum += digit;
    shouldDouble = !shouldDouble;
  }

  return sum % 10 === 0;
}

File: ios/Luhn.swift
import Foundation

final class Luhn {
  func isValid(_ pan: String) -> Bool {
    let digits = pan.compactMap { $0.wholeNumberValue }

    guard digits.count >= 12 else {
      return false
    }

    var sum = 0
    let reversed = digits.reversed()

    for (index, digit) in reversed.enumerated() {
      if index % 2 == 1 {
        let doubled = digit * 2
        sum += doubled > 9 ? doubled - 9 : doubled
      } else {
        sum += digit
      }
    }

    return sum % 10 == 0
  }
}
